### PR TITLE
Hide associated constants too when collapsing implementation

### LINKED ISSUE #71849: Rustdoc: strange collapse behaviour when impl block has comment
```rust
pub struct S;
impl S {
    pub const C1: () = ();
    pub fn foo1() {}
}
/// Comment
impl S {
    pub const C2: () = ();
    pub fn foo2() {}
}
/// Comment
impl S {
    pub const C3: () = ();
    pub fn foo3() {}
}
```
On the left of the image is the documentation for the three blocks.
1. No comment for the impl block.
2. The comment for the impl block is expanded.
3. The comment for the impl block is collapsed.

On the right, there is what happens when the three impl blocks are collapsed.
1. Collapsed as expected, with both constant and function hidden as expected.
2. The comment is still visible though its `[-]` is hidden. The function is hidden as expected but the constant is not hidden.
3. The comment is hidden as expected. The function is hidden as expected but the constant is not hidden.

collapse

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #71849.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -2257,9 +2257,9 @@ function defocusSearchBar() {
 
         function implHider(addOrRemove, fullHide) {
             return function(n) {
-                var is_method = hasClass(n, "method") || fullHide;
-                if (is_method || hasClass(n, "type")) {
-                    if (is_method === true) {
+                var should_hide = hasClass(n, "method") || hasClass(n, "associatedconstant");
+                if (should_hide === true || fullHide === true || hasClass(n, "type")) {
+                    if (should_hide === true || fullHide === true) {
```
Comment (pickfire) [inline src/librustdoc/html/static/main.js:2262]: I think keeping javascript naming would be more preferable.

```suggestion
                var shouldHide = hasClass(n, "method") || hasClass(n, "associatedconstant");
                if (shouldHide === true || fullHide === true || hasClass(n, "type")) {
                    if (shouldHide === true || fullHide === true) {
```

**Comment by GuillaumeGomez** [comment]
@pickfire The force of habit. ;)

**Review by pickfire — APPROVED** [review]
Looks good to me.

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -2257,9 +2257,9 @@ function defocusSearchBar() {
 
         function implHider(addOrRemove, fullHide) {
             return function(n) {
-                var is_method = hasClass(n, "method") || fullHide;
-                if (is_method || hasClass(n, "type")) {
-                    if (is_method === true) {
+                var shouldHide = hasClass(n, "method") || hasClass(n, "associatedconstant");
+                if (shouldHide === true || fullHide === true || hasClass(n, "type")) {
+                    if (shouldHide === true || fullHide === true) {
```
Comment (jyn514) [inline src/librustdoc/html/static/main.js:2262]: Why did you move out the check to `fullHide` here? There's no time you'd want to treat it separately from the method checks, right?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2262]: Well, in the case it's an associated constant. ;)

Comment (jyn514) [reply src/librustdoc/html/static/main.js:2262]: I don't understand what this means. I'm asking why `shouldHide` is treated separately from `fullHide`.

**Comment by jyn514** [comment]
@bors r-

I would like to get an answer to https://github.com/rust-lang/rust/pull/79796#discussion_r538792911 before this merges (even if it's just explaining why the change happened).

File: src/librustdoc/html/static/main.js (hunk 2 above)
Comment (pickfire) [reply src/librustdoc/html/static/main.js:2262]: I did noticed this part but both looks the same to me, either way looks fine but I prefer `fullHide` being part of `shouldHide`.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2262]: It's a bit different: we don't hide "type" elements, only their doc block. So the first condition is either we should completely hide it or its doc block, the second one is if we should hide the element only.

Comment (jyn514) [reply src/librustdoc/html/static/main.js:2262]: @GuillaumeGomez that's not my question. My question is why did you change from

```rust
var shouldHide = hasClass(n, "method") || hasClass(n, "associatedconstant") || fullHide; 
if (shouldHide) { ... }
```
to
```rust
var shouldHide = hasClass(n, "method") || hasClass(n, "associatedconstant"); 
if (shouldHide || fullHide) { ... }
```
The second is much easier to forget the check to `fullHide`, and there's no need to make it separate.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2262]: I thought on the contrary that it made the code easier to follow but I can put back `fullHide` into the variable.

Comment (jyn514) [reply src/librustdoc/html/static/main.js:2262]: That would be great, thanks :)

**Comment by GuillaumeGomez** [comment]
@jyn514 I have put back `fullHide` into the variable.

**Comment by GuillaumeGomez** [comment]
Forgot to r+ it...

@bors: r=jyn514 rollup
